pkg/operator/yugabytedb/create_controller.go

```diff
@@ -0,0 +1,343 @@
+
+	if _, err := c.context.Clientset.CoreV1().Services(cluster.namespace).Create(headlessService); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return err
+		}
+		logger.Infof("headless service %s already exists in namespace %s", headlessService.Name, headlessService.Namespace)
+	} else {
+		logger.Infof("headless service %s started in namespace %s", headlessService.Name, headlessService.Namespace)
+	}
+
+	return nil
+}
+
+func (c *ClusterController) createMasterStatefulset(cluster *cluster) error {
+	return c.createStatefulSet(cluster, false)
+}
+
+func (c *ClusterController) createTServerStatefulset(cluster *cluster) error {
+	return c.createStatefulSet(cluster, true)
+}
+
+func (c *ClusterController) createStatefulSet(cluster *cluster, isTServerStatefulset bool) error {
+	replicas := int32(cluster.spec.Master.Replicas)
+	name := masterName
+	label := masterName
+	serviceName := masterNamePlural
+	volumeClaimTemplates := []v1.PersistentVolumeClaim{
+		cluster.spec.Master.VolumeClaimTemplate,
+	}
+
+	if isTServerStatefulset {
+		replicas = int32(cluster.spec.TServer.Replicas)
+		name = tserverName
+		label = tserverName
+		serviceName = tserverNamePlural
+		volumeClaimTemplates = []v1.PersistentVolumeClaim{
+			cluster.spec.TServer.VolumeClaimTemplate,
+		}
+	}
+
+	// Append CR name suffix to make the service name & label unique in current namespace.
+	name = cluster.addCRNameSuffix(name)
+	label = cluster.addCRNameSuffix(label)
+	serviceName = cluster.addCRNameSuffix(serviceName)
+	volumeClaimTemplates[0].Name = cluster.addCRNameSuffix(volumeClaimTemplates[0].Name)
+
+	statefulSet := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: cluster.namespace,
+			Labels:    createAppLabels(label),
+		},
+		Spec: appsv1.StatefulSetSpec{
+			ServiceName:         serviceName,
+			PodManagementPolicy: appsv1.ParallelPodManagement,
+			Replicas:            &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: createAppLabels(label),
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: cluster.namespace,
+					Labels:    createAppLabels(label),
+				},
+				Spec: createPodSpec(cluster, c.containerImage, isTServerStatefulset, name, serviceName),
+			},
+			UpdateStrategy: appsv1.StatefulSetUpdateStrategy{
+				Type: appsv1.RollingUpdateStatefulSetStrategyType,
+			},
+			VolumeClaimTemplates: volumeClaimTemplates,
+		},
+	}
+	cluster.annotations.ApplyToObjectMeta(&statefulSet.Spec.Template.ObjectMeta)
+	cluster.annotations.ApplyToObjectMeta(&statefulSet.ObjectMeta)
+	k8sutil.SetOwnerRef(&statefulSet.ObjectMeta, &cluster.ownerRef)
+
+	if _, err := c.context.Clientset.AppsV1().StatefulSets(cluster.namespace).Create(statefulSet); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return err
+		}
+		logger.Infof("stateful set %s already exists in namespace %s", statefulSet.Name, statefulSet.Namespace)
```

Comment:
Kind of same here. Question for the future, should the StatefulSet and other components be "replaced" / "updated" when they already exist?

Having the Create ("start") logic as close as possible to be like the update logic would allow to "save" some lines and with that potential "bug" sources between both logics. ;)

Reply:
Agree. We plan to bring both logics closer in the next release.